Reinstate Windows CI Pipeline usefulness

TBBle opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

* Fixed false-green state from the Windows CI pipeline when unit tests failed.
* * This should apply to integration tests too.
* Fixed failing unit tests that [snuck into master](https://github.com/moby/moby/pull/41458#issuecomment-722306672) due to the false-green CI results.

Supplants: #41637

**- How I did it**

* Removed attempt to save and restore `$LastExitCode` in the `Finally` block: that block does not see the `$LastExitCode` for the exit that triggered it, and failures inside the block do not change the `$LastExitCode` seen after the block exits.
* * Separately tested [on CI](https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-41636/4/pipeline) to show that it fails without the next fix.
* * Also _with_ the next fix, [a failing Integration test](https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-41636/5/pipeline/252) will report failure correctly too.
* * Helpfully the next fix also left Integration CLI tests [breaking correctly and failing the build](https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-41638/2/pipeline/242).
* Split a small collection of non-Windows mount-related code and constants into __unix.go_ source files so Windows wouldn't see them.

**- How to verify it**

1. Break a test
2. Find out

**- Description for the changelog**

N/A, fixing a recent build break.

**- A picture of a cute animal (not mandatory but encouraged)**

![](https://i.imgur.com/D60BcMu.jpeg)

(Picture by https://twitter.com/mofu_sand)

TBBle committed:
Don't clear the exit code from the Finally block

It turns out that the Finally block does not see the exit code from the
`exit` call that triggered it, but from an earlier state. And it seems
that actions take in the Finally block other than `exit` will not affect
the $LastErrorCode set by the `exit` that triggered the Finally block.

Signed-off-by: Paul "TBBle" Hampson <Paul.Hampson@Pobox.com>

TBBle committed:
Separate non-Windows mount code from common code

Signed-off-by: Paul "TBBle" Hampson <Paul.Hampson@Pobox.com>

thaJeztah commented on hack/ci/windows.ps1:
```diff
@@ -1012,9 +1012,6 @@ Catch [Exception] {
-    # Preserve the LastExitCode of the tests
-    $tmpLastExitCode = $LastExitCode
-
     $ErrorActionPreference="SilentlyContinue"
```
I guess, the `Catch [Exception]` already should handle any error that occurred so far, and the `SilentlyContinue` won't mask those errors, correct?

TBBle replied:
Inside the `Try` everything is set up to Throw as a fail-out case, and then the `Catch` logs and it either turns it into an error code with `exit` for CI, or `Throw`s onwards for interactive use. As far as I could tell by testing (I couldn't find any good docs on this) `Finally` can't interfere with the `exit` result unless it also calls `exit`.

(I suspect `exit` is internally a `Throw` with a specific type caught at the top-level, based on the above, somewhat like Python's `sys.exit()`.)

I assume the `SilentlyContinue` here is because we don't want the `Finally` block to change the test outcome by throwing when we otherwise passed (most of the things in this block also explicitly `SilentlyContinue` anyway, like the log-collection) since for CI, most of the things we're doing here are not interesting (cleanups) and for interactive use, you wouldn't want a cleanup failure to overwrite a test failure result.

It might make sense to use `Continue` instead of `SilentlyContinue` so that failures in this block log their error messages, perhaps.

TBBle committed:
Move tests using sys/mount to not build on Windows

Signed-off-by: Paul "TBBle" Hampson <Paul.Hampson@Pobox.com>

TBBle committed:
Break out non-Windows sys/mount usage into helper

Rather than bifurcate the test completely, this lets us keep the test
intact with a small function wrapper to allow the compiler to build the
code that'll never be called on Windows, on Windows.

Signed-off-by: Paul "TBBle" Hampson <Paul.Hampson@Pobox.com>

TBBle commented:
Dang. Because I touched integration-cli (moved some tests) CI is seeing them as new tests and [rejects validation](https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-41638/3/pipeline/195). Also, the Windows pipeline failed for [network reasons](https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-41638/3/pipeline/237), I think.

I'm not sure how to bypass the former, and a new CI run should fix the latter. The [previous Windows run](https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-41638/2/pipeline/242) got through to the integration-cli tests okay, so I'm pretty confident the unit tests and integration tests are fine now.

thaJeztah commented:
We should be ok to ignore the RS1 failures (RS1 is not running by default, I triggered it manually to see if there's no huge problem), and doesn't look like a regression caused by this PR;

https://ci-next.docker.com/public/blue/rest/organizations/jenkins/pipelines/moby/branches/PR-41638/runs/4/nodes/316/log/?start=0

```
--- FAIL: TestBuildWithRemoveAndForceRemove (0.02s)
    --- PASS: TestBuildWithRemoveAndForceRemove/failed_build_with_no_removal (44.44s)
    --- FAIL: TestBuildWithRemoveAndForceRemove/successful_build_with_remove (44.77s)
        build_test.go:113: assertion failed: 0 (c.numberOfIntermediateContainers int) != 1 (int): Expected 0 remaining intermediate containers, got 1
    --- FAIL: TestBuildWithRemoveAndForceRemove/successful_build_with_remove_and_force_remove (45.41s)
        build_test.go:113: assertion failed: 0 (c.numberOfIntermediateContainers int) != 1 (int): Expected 0 remaining intermediate containers, got 1
    --- FAIL: TestBuildWithRemoveAndForceRemove/successful_build_with_no_removal (45.41s)
        build_test.go:113: assertion failed: 0 (c.numberOfIntermediateContainers int) != 1 (int): Expected 0 remaining intermediate containers, got 1
    --- PASS: TestBuildWithRemoveAndForceRemove/failed_build_with_remove (26.78s)
    --- PASS: TestBuildWithRemoveAndForceRemove/failed_build_with_remove_and_force_remove (26.59s)
```

Same test passes on RS5; https://ci-next.docker.com/public/blue/rest/organizations/jenkins/pipelines/moby/branches/PR-41638/runs/4/nodes/286/log/?start=0

```
--- PASS: TestBuildWithRemoveAndForceRemove (0.02s)
    --- PASS: TestBuildWithRemoveAndForceRemove/failed_build_with_no_removal (17.32s)
    --- PASS: TestBuildWithRemoveAndForceRemove/successful_build_with_no_removal (17.68s)
    --- PASS: TestBuildWithRemoveAndForceRemove/failed_build_with_remove_and_force_remove (17.93s)
    --- PASS: TestBuildWithRemoveAndForceRemove/failed_build_with_remove (17.94s)
    --- PASS: TestBuildWithRemoveAndForceRemove/successful_build_with_remove_and_force_remove (9.07s)
    --- PASS: TestBuildWithRemoveAndForceRemove/successful_build_with_remove (8.88s)
```

Unit tests passed on both RS1 and RS5;

```
DONE 1768 tests, 45 skipped in 287.554s
INFO: make.ps1 ended at 11/05/2020 19:45:45
INFO: Unit tests results(bundles\junit-report-unit-tests.xml) exist. pwd=D:\gopath\src\github.com\docker\docker
INFO: Unit tests ended at 11/05/2020 19:45:48. Duration:00:05:12.1634473
INFO: Building busybox
Sending build context to Docker daemon   5.12kB


DONE 1768 tests, 45 skipped in 236.483s
INFO: make.ps1 ended at 11/05/2020 19:35:13
INFO: Unit tests results(bundles\junit-report-unit-tests.xml) exist. pwd=D:\gopath\src\github.com\docker\docker
INFO: Unit tests ended at 11/05/2020 19:35:15. Duration:00:04:16.1988449
INFO: Building busybox
Sending build context to Docker daemon   5.12kB
```

And integration tests on RS5;

```
DONE 1094 tests, 547 skipped in 2761.701s
INFO: Integration tests ended at 11/05/2020 20:28:27. Duration:00:46:01.7842500
INFO: Docker info of the daemon under test at end of run
```

I manually ran the `docker-py` tests;

```
= 375 passed, 5 skipped, 1 deselected, 4 xfailed, 4 xpassed in 484.38 seconds ==
---> Making bundle: .integration-daemon-stop (in bundles/test-docker-py)
```

And cross;

```console
root@3b8fca845230:/go/src/github.com/docker/docker# hack/make.sh cross

Removing bundles/

---> Making bundle: cross (in bundles/cross)
Cross building: bundles/cross/linux/amd64
Building: bundles/cross/linux/amd64/dockerd-dev
GOOS="linux" GOARCH="amd64" GOARM=""
Created binary: bundles/cross/linux/amd64/dockerd-dev
Copying nested executables into bundles/cross/linux/amd64
Cross building: bundles/cross/windows/amd64
Building: bundles/cross/windows/amd64/dockerd-dev.exe
GOOS="windows" GOARCH="amd64" GOARM=""
Created binary: bundles/cross/windows/amd64/dockerd-dev.exe
Cloning into '/go/src/github.com/docker/windows-container-utility'...
remote: Enumerating objects: 2, done.
remote: Counting objects: 100% (2/2), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 23 (delta 0), reused 0 (delta 0), pack-reused 21
Unpacking objects: 100% (23/23), done.
Building: bundles/cross/windows/amd64/containerutility.exe
```

thaJeztah commented:
> Dang. Because I touched integration-cli (moved some tests) CI is seeing them as new tests and rejects validation.

So what we did on previous occasions, is to push a temporary commit to disable that check, and revert afterwards in a follow-up PR.


See my above comment; I tried the important tests that were skipped locally, and I'd be comfortable merging as-is; let me run CI once more to see if the RS1 failure is something to look into, or just some race condition (as RS1 is generally slower)

TBBle commented:
I'm happy to push as-is. Once this lands, I'll be rebasing #41636 onto it, and if any later validation errors do show up, I will see them there and follow up ASAP.

thaJeztah commented:
Currently, Docker doesn't publish Windows Server packages (they used to be part of Docker Enterprise), only Docker Desktop (which I _know_ some people install on Windows Server); bringing back Windows packages is being discussed, but no ETA yet. We _might_ drop Windows 2016 ("rs1") at that point as well (or at least; it wouldn't be a main point of focus), as it's ending it's shelve-life anyway, and overall  experience is so much worse than newer versions of Windows.

cpuguy83 commented:
Or we could move the tests to integration/ 😄 
Or just get rid of the check on integration-cli altogether.

thaJeztah commented:
I think the check is useful if we don't want the integration-cli to be added to(and we _did_ miss that on occasion); it should be easier to override though.

Moving the test would be ok of course; point is that currently, master is broken, and I'd love to see that fixed (and not put the burden of additional work on @TBBle)

TBBle commented:
I am in no way committing to do this, but for my general understanding, it looks like those two tests would be best moved to https://github.com/moby/moby/blob/master/integration/container/mounts_linux_test.go?

TBBle committed:
Rewrite VolumesMountedAsShared/Slave as Integration tests

This moves the two tests from integration-CLI to integration.

Signed-off-by: Paul "TBBle" Hampson <Paul.Hampson@Pobox.com>

cpuguy83 commented on integration/container/mounts_linux_test.go:
```diff
@@ -266,3 +267,127 @@ func TestContainerBindMountNonRecursive(t *testing.T) {
+
+	// Convert this directory into a shared mount point so that we do
+	// not rely on propagation properties of parent mount.
+	if err := mount.Mount(tmpDir1.Path(), tmpDir1.Path(), "none", "bind,private"); err != nil {
```
Nit, on these binds you can use `mount.MakeShared` or `mount.MakeSlave`. This takes care of self-binding as well as setting the propagation.

TBBle replied:
Oh, neat! I'll slip that improvement into #41636. Thank you.

cpuguy83 reviewed (approved):
Minor nit, but LGTM.

No idea about that windows.ps1, though.

thaJeztah reviewed (approved):
LGTM
